### ISSUE TITLE
Reduce Mangrove contract size below limit

### ISSUE THREAD

**Issue by adhusson** [mangrovedao/mangrove-core#470]
For now, Mangrove's contract size is too big for the ~24kB limit.

1. One way to reducing it would be to move all governance functions outside and just have a generic "modify config" function inside Mangrove.
2. Another would be to move all view functions outside and have a generic `execute arbitrary code` function inside Mangrove.

For the second point, the idea is: have Mangrove implement a function that runs arbitrary code but then reverts. So you can do anything, revert, and return the data you got. In pseudocode):

```
// in Mangrove
function run_code(address implem, bytes call_data) {
  (success, revert_data) = delegatecall(address(this),abi.encodeCall(this.run_code_and_revert,(implem,call_data)));
}

function run_code_and_revert(address implem, bytes call_data) {
  (success,return_data) = delegatecall(implem,data);
  revert(return_data);
}
```

Now if you have a MangroveViewFunctions contract with the same storage layout as Mangrove at address `0xViewFunctions`, you can do:

```
bytes memory view_call_data = abi.encodeCall(MangroveViewFunctions.viewSomething,(...args));
(success, view_data) = staticall(address(mgv),abi.encodeCall(Mangrove.run_code,(mgv_view_functions,view_call_data)));
```

**Comment by adhusson** [comment]
Writing down pros/cons of 2 approaches to reducing bytecode size by moving view functions outside of Mangrove

**Option A: Delegatecall to a contract with same storage layout**
Description:
* When mgv.best(...) is called, the fallback function is triggered and mangrove does `delegatecall(other_contract_with_same_storage_layout_as_mangrove,best()) `.

Variations:
* Allow calling arbitrary contracts (eg `user_supplied_contract_with_same_storage_layout_as_mangrove`) + revert before returning data so everything done in the call is cancelled (for safety)

Advantages:
* Direct call to mangrove for most view functions

Disadvantages:
* Deployment more complex (must deploy a third contract), a little hard to understand

**Option B: MgvReader implements view functions + Mangrove exposes a “read arbitrary storage slot” function**
Description
* Mangrove exposes a view extsloads function that takes an array of slots and returns an array of bytes32
MgvReader has function best() which finds out the correct storage slot to read and calls 
```solidity
extsloads([olKey_leaf])[0].getNextOfferId()
```

Advantages:
* No fallback-based magic in Mangrove

Disadvantage:
* Code for figuring out correct storage slots manually is ugly and contains hand-written slot values

**Comment by adhusson** [comment]
Example of code change with option **B**: 

```solidity
// (simplified)
// From this:
function _config(OLKey memory olKey) internal view returns (Global _global, Local _local) {
  offerList = offerLists[olKey.hash()];
  _global = internal_global;
  _local = offerList.local;
  // ...
}

// To this:
function local_slot(bytes olKeyHash) internal pure returns (bytes32 slot) {
  OfferList storage offerList = offerLists[olKeyHash]
  assembly("memory-safe") {
    slot := offerList.slot + 0 // `local` has slot 0 within an OfferList struct
  }
}

function _config(OLKey memory olKey) internal view returns (Global _global, Local _local) {
  bytes res;
  res = mgv.extsload(GLOBAL_SLOT) // GLOBAL_SLOT is an immutable
  assembly {
    _global := res
  }
  bytes32 olKeyHash = olKey.hash();
  res = mgv.extsload(local_slot(olKeyHash));
  assembly {
    _local := res
  }
  // ...
}
```

**Comment by adhusson** [comment]
(Ended up choosing option A)

**Comment by adhusson** [comment]
Closed by #499
